### DISCUSSION TITLE
Maximum call stack size exceeded when I add in a validation function?

### DISCUSSION THREAD

**Discussion by durianwaffle** [eclipse-langium/langium#1670] [Q&A] [answered]
I wrote my own validator, and registered it in `my-module,ts file` like `registerValidationChecks(My);`
But after I added the checks, I found that **parsing autocompletion, and tooltip failed**. There is a **rangeError: Maximum call stack size exceeded** in `MyAstReflection.computeIsSubtype src/language/generated/ast.ts:812:29`.
```
src/language/generated/ast.ts:812:29
    810|             }
    811|             case NamedElement: {
    812|                 return this.isSubtype(StElement, supertype) || this.isSubtype(VariableDeclaration, supertype);
       |                             ^
    813|             }
    814|             case Statement: {
```
My code about the validator is as below:

```{TypeScript}
import type { ValidationAcceptor, ValidationChecks } from 'langium';
import type { AssignmentStatement, MyAstType } from './generated/ast.js';
import { VariableReference } from './generated/ast.js';
import type { MyServices } from './structured-text-module.js';

/**
 * Register custom validation checks.
 */
export function registerValidationChecks(services: MyServices ) {
    const registry = services.validation.ValidationRegistry;
    const validator = services.validation.StructuredTextValidator;
    const checks: ValidationChecks<MyAstType > = {
        AssignmentStatement: validator.checkAssignmentType
    };
    registry.register(checks, validator);
}

/**
 * Implementation of custom validations.
 */
export class MyValidator {
    checkAssignmentType(assignemntStatement: AssignmentStatement, accept: ValidationAcceptor): void {
        if (assignemntStatement.left) {
            const left = assignemntStatement.left as VariableReference;
            const leftType = left.$type;
            const rightType = 'BitVariable';
            console.log('Left op of assignment', leftType);
            if (leftType.toLocaleUpperCase() !== rightType) {
                accept('error', `Can’t assign ${leftType} type to variable with type ${rightType} assignment.`, { node: assignemntStatement, property: 'left' });
            }
        }
    }
}
```

**Comment by msujew** [comment]
Hey @durianwaffle,

Can you share a reproducible example? It's hard to tell what's wrong from the snippet you provided.

**Reply by durianwaffle** [reply]
Hi @msujew , I added a console log in AstReflection, and found that it says computeIsSubtype called with subtype: `VariableDeclaration`, supertype: `BinaryExpression` before it crashed.

![image](https://github.com/user-attachments/assets/ec7ddc36-9942-4da7-8db0-b4c6a344d22b)

https://github.com/durianwaffle/validation-dsl/blob/main/examples/maxiumcallstacksizeexceed/README.md

**Reply by msujew** [reply]
I don't know what you did with the git submodule, but I can't seem to open it:

https://github.com/durianwaffle/Langium-Validation-Subtype-Call-Stack-Overflow/blob/main/examples/maxiumcallstacksizeexceed

![grafik](https://github.com/user-attachments/assets/eb7f19de-19b4-4558-826c-4b03735d40f9)

Running also isn't possible in git:

> $ git submodule init
fatal: No url found for submodule path 'examples/maxiumcallstacksizeexceed' in .gitmodules

**Reply by durianwaffle** [reply]
Hi @msujew sorry for that!
I updated the link. Now you can see it.
![image](https://github.com/user-attachments/assets/42bd192a-d7b6-4aec-abbc-8c3b0558a5db)

https://github.com/durianwaffle/validation-dsl/blob/main/examples/maxiumcallstacksizeexceed/README.md

**Reply by msujew** [reply]
You've produced a cyclical inheritance chain that hasn't been discovered by our type system validation system.

Effectively, you've written your types like this:

```ts
type VariableDeclaration = NamedElement;
type NamedElement = VariableDeclaration;
```

To fix this, you need to remove the `returns NamedElement` from all the parser rules.

**Reply by durianwaffle** [reply]
`NamedElement` is there to make membercalls. 
In the LOX sample:   "VariableDeclaration returns NamedElement"
But there is no "maximum call stack size" issue in LOX sample.
Why is it?

**Reply by msujew** [reply, marked as answer]
No idea why it works in Lox, it's likely a very specific set of circumstances. It's not supposed to work like this. You either write:

```
type NamedElement = VariableDeclaration | ...;
```

or

```
VariableDeclaration returns NamedElement: ...;
```

but not both.

**Reply by durianwaffle** [reply]
Hi @msujew, a follow up question.
After I deleted the other returns NamedElement, the validator works.
But I will see `Coud not resolve reference to NamedElement` when I **access struct within a struct**.
Do you know how I can troubleshot for this issue?
![image](https://github.com/user-attachments/assets/c399fa40-005e-4938-a7d9-ede10093ff31)

**Reply by msujew** [reply]
@durianwaffle Looks like a scoping issue. You should look into your scoping/type system code to ensure everything works as expected there.

**Reply by durianwaffle** [reply]
Okay thank you @msujew ! It helps a lot!
